Export collections when exporting an history

FredericBGA opened the pull request:
#4345 

Here are some first commits.

Collections are exported and their attributes and files are now present in the archive.

I have to deal now with import. I hope that I have exported enough stuff within the archive

Exports works within my install (I've add a column in one table of the database: job_export_history_archive). 
It's not heavily tested (two small histories, one with only one collection, another one with two collections)

Collections are type "list", I haven't tried type "pair" yet

I used pep8 in order to avoid common errors of style

I don't know what to do here: 

> lib/galaxy/model/migrate/versions/

FredericBGA committed:
collections are now exported with history, archive contains everything needed for import

FredericBGA committed:
add collections_attrs_file to enable collections export

FredericBGA committed:
W291 trailing whitespace

FredericBGA committed:
pep8 --first corrections

nsoranzo commented on lib/galaxy/model/mapping.py:
```diff
@@ -582,7 +582,8 @@
     Column("history_attrs_filename", TEXT),
     Column("datasets_attrs_filename", TEXT),
-    Column("jobs_attrs_filename", TEXT))
+    Column("jobs_attrs_filename", TEXT,
```
You're missing a `)` between `TEXT` and `,`

nsoranzo commented:
> I used pep8 in order to avoid common errors of style

What I normally do is:
```
# Setup (only first time)
[galaxy]$ virtualenv .venv
[galaxy]$ . .venv/bin/activate
(.venv)[galaxy]$ pip install --upgrade pip setuptools
(.venv)[galaxy]$ pip install --upgrade flake8 tox
# Testing
[galaxy]$ . .venv/bin/activate
(.venv)[galaxy]$ tox -e py27-lint
```
You also can run locally other Continuous Integration testing by substituting `py27-lint` with `py27-lint-imports-include-list`, `py34-lint`, `py27-unit`, `qunit` or `first_startup` (you don't need to run them all, just the ones that may be affected but your local changes).

FredericBGA committed:
missing a ) between TEXT and ,

FredericBGA commented:
I don't know how to get the collection_type?
Could you help me?
I will need it for the import process.

Do I need to add something here:
[https://docs.galaxyproject.org/en/master/_modules/galaxy/model.html#DatasetCollectionInstance.state](url)

jmchilton commented on lib/galaxy/tools/imp_exp/__init__.py:
```diff
@@ -395,6 +406,27 @@ def default(self, obj):
+                if isinstance(obj, trans.model.HistoryDatasetCollectionAssociation):
+                    # dump attrs of each dataset of this collection
+                    collections_datasets_attrs = []
+                    collection_datasets = obj.dataset_instances
```
This really flattens out the dataset collection - so you loose all collection nesting informations and element identifiers that are very important I think. You want to somehow recursively encode ``object.collection`` - the ``DatasetCollection`` object. This will include the ``collection_type`` and element identifiers.

FredericBGA replied:
ok, I've got collection_type now. Thank you, I've missed the fact that I can acces the collection object itself!
But I'm not sure I understand well this part:

> You want to somehow recursively encode object.collection - the DatasetCollection object. This will include the collection_type and element identifiers.

I've tried, I retrieve indeed `DatasetCollection` object. `DatasetCollection` gives me a lot of elements (state, populated) and objects (dataset_instances and dataset_elements) but what do you think I will need to export?
There is already a lot of data in my file:
[collections_attrs.txt](https://github.com/galaxyproject/galaxy/files/1269883/collections_attrs.txt)

The related dataset files are stored in a directory (collections_datasets, maybe it could fit in the already created datasets directory but it helps me right now to see if I have of the files I need)

jmchilton commented:
This is a really wonderful start - thanks so much for taking this on! Let me know if you get stuck and I can try to find some time to refine it to get that nesting information right - but you seem to have made incredible progress so far.

FredericBGA committed:
add collection_type, clean code

FredericBGA commented:
@jmchilton thank you, I will try now the import part.

Sure I will need some insights!

Maybe the way you see the process:

- create the collection

- create the datasets (from collections) or create them the same way as the regular datasets

- link datasets to collection

and some hints related to objects that will be usefull for this.

FredericBGA commented:
@jmchilton: I hope I can do something like this (line 121 of lib/galaxy/tools/imp_exp/__init__.py):
`datasets_attrs += provenance_attrs`

and then add dataset(s) from collection(s):
```
    collections_attrs_file_name = 'collections_attrs.txt'
    collections_attr_str = read_file_contents(collections_attrs_file_name)
    collections_attrs = loads(collections_attr_str)
    
    for collection_attrs in collections_attrs:
        # add datasets attrs from collection to the global datasets_attrs
        datasets_attrs += collection_attrs['datasets']
        # create collection
        # create a dict ? coll/dataset dataset/coll in order to link them when creating datasets
```

All the datasets creation stuff will be then already handled, I hope it will work because I've used the same `HistoryDatasetAssociationEncoder` so I can retrieve everything needed in my file `collections_attrs.txt`

I will have to find how to link dataset(s) and collection(s). I will dig into this. But it will be maybe faster if you point me the right objects.

FredericBGA committed:
export and import elements, import collection works when datasets are not hidden

FredericBGA commented:
few updates.
exports of elements works.
import of collection (and their datasets) works _only_ when the datasets are not hidden in history.
I will deal with this issue: create datasets that are in a collection, but not created yet because they have the `exported `flag set to false

FredericBGA committed:
import collection works when datasets are hidden

FredericBGA commented:
The last thing to do is to export jobs attributes for collections.
it seems to be straigthforward.

FredericBGA committed:
export jobs attributes for dataset in collection

FredericBGA commented:
seems to be ok.
At least it works as I expected, collections are now fully exported.

I see that `Size on Disk` is not exported. But that was already the case before this PR. I don't know why.

I will maybe try to be a bit smarter and not export collection_datasets when these datasets are already exported.

jmchilton commented:
Does it work for collections that are lists of pairs? I just don't see how that is going to work without some level recursion. If it doesn't work for that but it does work for flat collections (lists and pairs) - lets at least make sure we are outputing formats that will allow us to do nesting in a backward/forward compatible way in the future. I'll try to find some more time to take another pass at this - thanks for the continued progress - it is looking great!

FredericBGA commented:
No, it doesn't (export doesn't fail, but the imported history is broken)
It works only like you said for flat collections (lists and pairs).
Lists of pairs were not one of my use case, so I didn't work on that.

jmchilton committed:
Simplify path tracking related to history exports.

This is a simpler model to think about and prevents us from needing to track new fields in the database for extensions such as collections.

jmchilton committed:
Rework export history argument parsing.

Other changes in #4532 would have caused export jobs that were created before an upgrade that added that PR to fail after the upgrade. This wouldn't have been a huge deal but the new format of the arguments is a bit cleaner and future proof anyway.

jmchilton commented:
Pushed some commits into this branch to simplify the import/export process and fix some of the changes in this PR.

I still have some issues I'd like to see resolved before merging. I know they are outside your scope - so I'll try to work on them. 

- [ ] Handle nested collections in collection attrs.
- [ ] Current the dataset <-> collection mapping assumes all the datasets have HIDs that appear elsewhere in the history - but a dataset doesn't have to come from the same history. I have a vague notion that we should track both the datasets in a history and then the remaining non-history or hidden collection datasets in a separate file and connect the mapping based on either filename or the exporting Galaxy's ID.

I'll update this PR if I make progress on these.

mvdbeek commented:
Thanks a lot @FredericBGA for getting the ball rolling, the export and import of collections will be included in 19.05!

FredericBGA commented:
> 
> 
> Thanks a lot @FredericBGA for getting the ball rolling, the export and import of collections will be included in 19.05!

you're welcome, thank you and @jmchilton
